c++/src/capnp/rpc.c++

```diff
@@ -1672,43 +1663,46 @@ private:
-                .described = described,
-              };
-            } else  {
-              // writeThirdPartyDescriptor() ended up deciding that it's not pointing at a
-              // third party after all, oops. This is unusual, but DeferredThirdPartyClient has
-              // some edge cases where this happens.
-
-              // Unfortunately, this will leave a hole in the message. This should be rare.
-              // TODO(perf): Since we're removing the last object allocated, the segment ought
-              //   to be able to reclaim the memory, but this doesn't appear to be implemented
-              //   at present. Fix that?
-              descriptor.disownThirdPartyHosted();
-              descriptor.setNone();
-
-              // Recursively apply to the client returned by `writeThirdPartyDescriptor()` (which
-              // is not actually a vine in this case).
-              return writeDescriptor(*result.vine, descriptor, fds);
+            KJ_SWITCH_ONEOF(result) {
+              KJ_CASE_ONEOF(vine, RpcClient::VineToExport) {
+                // We always have to add a new export to the table to use as the "vine"; we cannot
+                // share an existing export pointing to the same capability. The reason is, the
+                // vine represents not just the capability but *this specific handoff* of the
+                // capability; when it is dropped, we can clean up the handoff. This is
+                // accomplished by attaching the `vineInfo` to the export.
+                //
+                // Additionally, the vine is never treated as a promise, even if the capability
+                // backing it is a promise.
+                ExportId exportId;
+                auto& exp = exports.next(exportId);
+                exp.canonical = false;
+                exp.refcount = 1;
+                exp.clientHook = kj::mv(vine.vine);
+                exp.vineInfo = kj::mv(vine.vineInfo);
+
+                tph.setVineId(exportId);
+                return {
+                  .exportId = exportId,
+
+                  // The vine is the correct inner cap to use for disembargo purposes.
+                  .described = *exp.clientHook,
+                };
+              }
+              KJ_CASE_ONEOF(redirect, kj::Own<ClientHook>) {
+                // writeThirdPartyDescriptor() ended up deciding that it's not pointing at a
+                // third party after all, oops. This is unusual, but DeferredThirdPartyClient has
+                // some edge cases where this happens.
+
+                // Unfortunately, this will leave a hole in the message. This should be rare.
+                // TODO(perf): Since we're removing the last object allocated, the segment ought
+                //   to be able to reclaim the memory, but this doesn't appear to be implemented
+                //   at present. Fix that?
+                descriptor.disownThirdPartyHosted();
```

Comment:
This whole thing feels odd. Why doesn't the client hook call back into the connection when it knows what it wants to do, instead of the connection assuming what's going to happen?